x-pack/plugins/apm/server/lib/traces/get_top_traces.js

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { getTransactionList } from '../transactions/get_transaction_list';
+import { TRANSACTION_ID, TRACE_ID, PARENT_ID } from '../../../common/constants';
+import { get } from 'lodash';
+
+export async function getTopTraces({ setup }) {
+  const { start, end } = setup;
+  return getTransactionList({
+    setup,
+    query: {
+      bool: {
+        must: {
+          exists: {
+            field: TRACE_ID
+          }
+        },
+        must_not: {
+          exists: {
+            field: PARENT_ID
+          }
+        },
+        filter: [
+          {
+            range: {
+              '@timestamp': {
+                gte: start,
+                lte: end,
+                format: 'epoch_millis'
+              }
+            }
+          }
+        ]
+      }
+    },
+    transformResult: (bucket, { avg, tpm, impact, sample }) => ({
+      name: bucket.key,
+      serviceName: get(sample, 'context.service.name', 'n/a'),
+      id: get(sample, TRANSACTION_ID),
+      p95: bucket.p95.values['95.0'],
+      averageResponseTime: avg,
+      tracesPerMinute: tpm,
+      impact
+    })
+  });
```

Comment:
I think this level of abstraction makes the code harder to read. I'd prefer less DRY code which is more readable - especially when the duplication is only in two places.

Perhaps you can extract just the aggregation?

Reply:
Btw. I can totally related to this. When I first started in Elastic I wrote a DSL to get rid of the duplication in ES queries. I definitely went way further than you did :)
https://github.com/elastic/x-pack-kibana/pull/2653

Reply:
Yeah, I find this easier to understand and follow because it makes it clear what's different and what's not and helps confirm that traces are transactions, but I know abstraction has a trade-off cost. I don't mind repeating things, but I'm worried we're introducing a fair amount of code to retrieve "traces" that is actually just getting transactions with slightly different criteria to query for. What I think I'd really prefer is removing the `transformResult` altogether and aligning the results from both queries to have the same keys, which would make this even more streamlined.

Reply:
I see. I think if you can get rid of `transformResult` it would make it a lot more readable, and I'd be good with it 👍